### PR TITLE
add maxFailoverCount limit to TiKV

### LINKED ISSUE #508: add maxFailoverCount to TiKV
## Feature Request

**Is your feature request related to a problem? Please describe:**

We should add a maxFailoverCount to TiKV just like [TiDB](https://github.com/pingcap/tidb-operator/blob/4c183422345d8fe9e5a450f3729339ee01684767/pkg/manager/member/tidb_failover.go#L46) does.

Or the operator will create another 3 TiKV instances if all the 3 TiKV instances down but The PD instances are health.

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### LINKED ISSUE #963: TiKV cluster not respecting CRD replica configuration
## Bug Report

**What version of Kubernetes are you using?**
1.11.10 on EKS

**What version of TiDB Operator are you using?**
1.0.1

**What storage classes exist in the Kubernetes cluster and what are used for PD/TiKV pods?**
gp2

**What's the status of the TiDB cluster pods?**
Running, although we did have the single PD in a crash loop due to OOM and we have increased its memory.

**What did you do?**
We had a 1 PD, 3 TiKV and 1 TiDB pod setup, which was working fine. The PD pod was killed for some reason and it got into a crash loop backoff due to OOM. Then the number of pods of the TiKV configuration was increased drastically to 45 over the course of 8 hours.

**What did you expect to see?**
I expected the controller to
1) not increase the number of TiKV pods while the CRD specified 3 when PD
2) restore to 3 replicas when the PD pod was back Running.

**What did you see instead?**
The cluster remain at 45 replicas and I believe the controller might have some internal state that is corrupt. When using tkctl, I get
```
$ ~/Downloads/tkctl info -t data-monitoring--tidb-cluster
Name:               data-monitoring--tidb-cluster
Namespace:          production
CreationTimestamp:  2019-09-25 10:08:59 -0700 PDT
Overview:
         Phase   Ready  Desired  CPU   Memory  Storage  Version
         -----   -----  -------  ---   ------  -------  -------
  PD:    Normal  1      1        100m  256Mi   1Gi      pingcap/pd:v3.0.1    
  TiKV:  Normal  45     45       250m  2Gi     20Gi     pingcap/tikv:v3.0.1  
  TiDB   Normal  1      1        250m  1Gi              pingcap/tidb:v3.0.1  
Endpoints(ClusterIP):
  Cluster IP:  172.20.124.214
```

but the CRD correctly points the desired is 3
```
$ k get tidbcluster
NAME                            PD                  STORAGE   READY   DESIRE   TIKV                  STORAGE   READY   DESIRE   TIDB                  READY   DESIRE
data-monitoring--tidb-cluster   pingcap/pd:v3.0.1   1Gi       1       1        pingcap/tikv:v3.0.1   20Gi      45      3        pingcap/tidb:v3.0.1   1       1
```

Restarting the controller pod did not help. Setting the replica set size to 3 directly made the controller bring it back to 8. Then setting it to 3 again made it stay at 3. However, the controller logs have things like
```
E0927 15:14:00.760483       1 tidb_cluster_controller.go:245] TidbCluster: production/data-monitoring--tidb-cluster, sync failed pod "data-monitoring--tidb-cluster-tikv-29" not found, requeuing
```
for each pod that is missing and they still appear as part of the CRD status.

**Supplementary material**
CRD status: https://pastebin.com/haQwqa6Q
Pod list: https://pastebin.com/1nGR98Lv

### PR DESCRIPTION
**Pull request by weekface**
### What problem does this PR solve? 

- This PR fixes #963, and closes #508 

we have added this limit to TiDB in this PR: https://github.com/pingcap/tidb-operator/pull/163 

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test

Code changes

 - Has Helm charts change
 - Has Go code change
 - Has documents change

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note
add `maxFailoverCount` limit to TiKV
 ```

@xiaojingchen @tennix @cofyc @onlymellb PTAL

### CODE REVIEW

File: pkg/manager/member/tikv_failover.go [hunk 1]
```diff
@@ -30,6 +31,13 @@ func NewTiKVFailover(tikvFailoverPeriod time.Duration) Failover {
 }
 
 func (tf *tikvFailover) Failover(tc *v1alpha1.TidbCluster) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+	if len(tc.Status.TiKV.FailureStores) >= int(tc.Spec.TiKV.MaxFailoverCount) {
+		glog.Errorf("%s/%s failure stores count reached the limit: %d", ns, tcName, tc.Spec.TiKV.MaxFailoverCount)
+		return nil
```
Comment (onlymellb) [inline pkg/manager/member/tikv_failover.go]: The return is nil, so is it more appropriate to output log with warning level?

File: charts/tidb-cluster/values.yaml [hunk 2]
```diff
@@ -261,6 +261,7 @@ tikv:
   # Specify the priorityClassName for TiKV Pod.
   # refer to https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/#how-to-use-priority-and-preemption
   priorityClassName: ""
+  maxFailoverCount: 3
```
Comment (DanielZhangQD) [inline charts/tidb-cluster/values.yaml]: Suggest to add comments here to make it clear what's the detailed meaning and maybe how to configure the value, e.g. for the cluster with 100 TiKV, what's the suggested config.

File: pkg/manager/member/tikv_failover.go [hunk 3]
```diff
@@ -30,6 +31,13 @@ func NewTiKVFailover(tikvFailoverPeriod time.Duration) Failover {
 }
 
 func (tf *tikvFailover) Failover(tc *v1alpha1.TidbCluster) error {
+	ns := tc.GetNamespace()
+	tcName := tc.GetName()
+	if len(tc.Status.TiKV.FailureStores) >= int(tc.Spec.TiKV.MaxFailoverCount) {
```
Comment (aylei) [inline pkg/manager/member/tikv_failover.go]: IIUC, If multiple tikv-servers become `Down` simultaneously, the `MaxFailoverCount` may be exceeded.

File: pkg/manager/member/tikv_failover.go (hunk 1 above)
Comment (onlymellb) [reply pkg/manager/member/tikv_failover.go]: IIRC, `tidb_failover.go` also needs to be modified.

File: pkg/manager/member/tikv_failover.go (hunk 3 above)
Comment (weekface) [reply pkg/manager/member/tikv_failover.go]: Good catch!

Addressed PTAL again.

File: charts/tidb-cluster/values.yaml (hunk 2 above)
Comment (weekface) [reply charts/tidb-cluster/values.yaml]: Done.

File: pkg/manager/member/tikv_failover.go [hunk 4]
```diff
@@ -47,6 +51,11 @@ func (tf *tikvFailover) Failover(tc *v1alpha1.TidbCluster) error {
 			if tc.Status.TiKV.FailureStores == nil {
 				tc.Status.TiKV.FailureStores = map[string]v1alpha1.TiKVFailureStore{}
 			}
+			if len(tc.Status.TiKV.FailureStores) >= int(tc.Spec.TiKV.MaxFailoverCount) {
+				glog.Warningf("%s/%s failure stores count reached the limit: %d", ns, tcName, tc.Spec.TiKV.MaxFailoverCount)
+				return nil
+			}
+
```
Comment (onlymellb) [inline pkg/manager/member/tikv_failover.go]: I think `tidb_failover.go` also needs to be modified like this

Comment (weekface) [reply pkg/manager/member/tikv_failover.go]: Done.

**Pull request merged**